### ISSUE TITLE
Retros support non-users

### ISSUE THREAD

**Issue by mattkrick** [ParabolInc/action#1722]
# Background
Some clients want contractors or new team members to be able to participate in a retro without signing up for parabol. This could potentially be detrimental to our bottom line, so it'll need to be helpful, yet leave something to be desired to encourage a user conversion. This is NOT part of the MVP.

# AC
- Anyone can join and participate in the retro if they have the link.
- Anonymous participants can be disabled in Team Settings (since this can clearly be a gaping privacy hole)

# Proposed Design:
- the `/retro/:teamId` route is open to the world
-A "Sign In" button is present in the upper right hand corner next to the avatars if the user is not logged in
- During the Discuss phase, non-users cannot create cards. The create card button will be grayed out and tell them to become a member.
- If there is no auth, a "retroUserId" is created & stored in the localStorage. This is superior to an IP address because turning a vpn on/off would be a standard practice. The ID is a composite key of an adjective & an animal (eg `HappyElephant`). It will be randomly generated & we'll ensure it doesn't already exist in the retro before it's written to the localStorage.
- An avatar will be uniquely generated client-side based on the name & will be concatenated to the members
- When the meeting has ended, non-users will be asked to provide their email to receive the summary. The have the option of declining, which will send them to the summary page.
- Their email will be persisted in a backend DB where we can sell them on the product through marketing emails
- If they log in mid-meeting, then the log-in mutation will include their temporary userId & source == RETRO.This will trigger a mutation to change the creatorId and votes for that anonymous user to a known user.

 

- **Estimated effort:** 12 points ([see CONTRIBUTING.md](https://github.com/ParabolInc/action/blob/master/CONTRIBUTING.md#points-and-sizes))

**Comment by ackernaut** [comment]
I really like the idea of how having a retro with anonymous users can help solve the chicken-and-egg content problem of our funnel. I think any user could create a card, but they can only assign it to the team lead who initiated the retro (who had to become a user, in my mind), or they use the option to assign by inline invitation via email. We could send a notification to the team lead if users don’t follow through with their invitations. Then we have content for the action meeting based on the cards that get generated from a retro! We could also follow up with the team lead to both schedule the first action meeting (sending them an email with a link to the “Getting Started Guide”). After 3-5 action meetings, we could prompt the team lead to hold another retro. (just brainstorming here)

**Comment by mattkrick** [comment]
i hear ya, we gotta make it easy to get in, but also draw the line somewhere to make them convert.
I'll argue that instead of letting anonymous people create cards, we allow team members to create cards for them a la soft team members. Here's why:
- The latter is already built, we can use soft team members today
- Someone with nothing more than the meeting url should not be able to invite someone to a team (inline invites on the assign menu). If i bring in a contractor for a retro, I value his input to the retro, but not the team flow. Why should he be allowed to invite a team member if he isn't a team member? Heck, he could even invite himself (although it'd still need approval). 
- it's a big job to handle anonymous Tasks & Invitations. Every client view & mutation on the server will have to be refactored to support that. Imagine the Team Settings page-- you have a bunch of outstanding invitations that were created by "Anonymous". Or a kanban with a Task created by "Anonymous". I'd argue that we need a little accountability (which also doubles as motivation for team members to invite the person). You can create a Task for a soft team member, but an anonymous person cannot create a task for you.

**Comment by ackernaut** [comment]
Good thoughts @mattkrick overall I just want it to be as smooth an entry as possible, be sticky, and have the right ways for teams to follow up after their initial retro. I’m sure we will dial in a nice V1.

**Comment by jordanh** [comment]
@mattkrick & I met this morning and discussed the following AC:

   - The meeting lobby is extended to include a special "meeting waiting list" table – it's a place where folks who haven't been approved to join the meeting hang out until they are approved to join the meeting (not the team! read on...)
   - When an un-authenticated user hits the meeting route, they see controls to Sign In/Create Account (it's probably floated as a modal)
      - At this point, we capture an email and password for the user – creating a Parabol account, MeetingParticipant, and SoftUser
      - We add this MeetingParticipant to the "meeting waiting list" and show controls for approving or rejecting them to every fully-fledged TeamMember
    - Any team member can approve the meeting participant to join the meeting
       - They can be approved from the lobby using the controls on the "meeting waiting list" table or,
       - If they join after the meeting has begun, the waiting participants shows up as special faded avatar(s) which when clicked allows any user to approve or reject them from joining the meeting (these avatars also show up in the Meeting Lobby)
         - If a Billing Leader approves them, three choices are possible (Approve to Team, copy informs about billing change), Add to Meeting Only, Reject
         - If a non-Billing Leader approves them, only Add to Meeting and Reject appear
         - If a non-Billing Leader approves them, their avatar gets a special badge that indicates they are a MeetingParticipant only and if a Billing Leader is present they still have the opportunity to Approve to Team before the meeting ends
    - A waitlisted MeetingParticipant sees a modal indicating that they are waiting for a team member to approve them
       - When approved they join the meeting, wherever the facilitator is navigated
       - If rejected they are told who rejected them, a continue button appears that when clicked will redirect them to `/me`
   - When a TeamMember rejects a user, the MeetingParticipant displays a count (a number of red Xs?) for the number of times they have been rejected
   - Once approved, the MeetingParticipant interacts in the meeting as usual
      - They can Facilitate (if promoted)
      - They can create and be assigned tasks
   - When the meeting concludes:
      - They receive the meeting summary
      - If they were not approved by a Billing Leader, the MeetingParticipant is treated as a SoftTeamMember and their tasks and team membership pend approval
        - Their team sidebar indicates that their team membership is pending org approval
          - If the SoftTeamMember clicks on this pending team, they see a modal explaining with copy explaining who they are waiting on and why
 
EE = 15 points
DU = 2

**Comment by jordanh** [comment]
^ when this design issue is completed, this is a Milestone...perhaps

**Comment by ackernaut** [comment]
Sketched up some concepts [in Figma](https://www.figma.com/file/XNxOV2fnHf67Ds2jjn4SzqmY/Retro-Guests?node-id=0%3A1)

**Comment by ackernaut** [comment]
Here’s the latest proposal:
- New retro meetings have a unique ID and URL
- Folks can sign in or create an account from the retro lobby
- Guests see a lobby asking for a 6 digit passcode to join (whether meeting is active or inactive)
- Team members can see a 6 digit passcode that can be shared with guest participants (along with the meeting URL)
- Guest avatars have a guest badge (suggesting Material’s `verified_user` icon)
- Optional: we add a 3rd tab under team settings called “Meetings”. There is an option that defaults to checked: “Guests can join retro meetings with a passcode” (If this is unchecked, guests cannot access the meeting if they are not on the team, and there is no passcode to share)
See [Figma concepts](https://www.figma.com/file/XNxOV2fnHf67Ds2jjn4SzqmY/Retro-Guests?node-id=18%3A62) for how this may shape up. We can refine details based on what we agree to implement.

**Comment by mattkrick** [comment]
Moving forward with the password idea:
- [ ] the password (6 digit number) is for access to the meeting, it's not attached to the user in any way
- [ ] the password should be shareable before the meeting has started to allow folks to email it ahead of time
- [ ] when hitting the meeting route & not authenticated, ask the user to create an account/login. tell them it's to send them the meeting summary
- [ ] when hitting the meeting route & not authorized, ask the user to enter a 6 digit password
- [ ] when hitting the meeting route & authorized, show the password so they can share it with others
- [ ] 1 meeting has 1 immutable password.

**Comment by ackernaut** [comment]
Re: password details — sounds good to me!

**Comment by mattkrick** [comment]
- so for each meeting type, we'll have to have a password that is on deck. it's created when the team is created & after each meeting conccludes.
- the url should be `/:meetingType/:teamId/` it should not include a password, that should be kept in internal state
- when a guest enters an incorrect password, we should say something like "123456 is not correct. Ask a team member to resend the password"
- when a guest enters a correct password, we'll have to add them as a soft team member
- soft team members should be able to see their "Soft Teams" on the sidebar. it should be the team name, but disabled. 
- during the meeting, the billing leader can approval the guest via dropdown. we tell them how much it'll cost.

**Comment by ackernaut** [comment]
Closing as next steps are captured in #2391
